flink-ml-lib/src/main/java/org/apache/flink/ml/classification/knn/KnnModelData.java

```diff
@@ -0,0 +1,165 @@
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.api.Schema;
+import org.apache.flink.types.Row;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+
+/** Knn model data, which stores the data used to calculate the distances between nodes. */
+public class KnnModelData {
+    private final List<Tuple3<DenseMatrix, DenseVector, int[]>> dictData;
+    private final Comparator<? super Tuple2<Double, Integer>> comparator;
+
+    /**
+     * Constructor.
+     *
+     * @param list Row list.
+     */
+    public KnnModelData(List<Row> list) {
+        this.dictData = new ArrayList<>(list.size());
+        for (Row row : list) {
+            this.dictData.add(
+                    Tuple3.of(
+                            (DenseMatrix) row.getField(0),
+                            (DenseVector) row.getField(1),
+                            (int[]) row.getField(2)));
+        }
+        comparator = Comparator.comparingDouble(o -> -o.f0);
+    }
+
+    /**
+     * Gets comparator.
+     *
+     * @return Comparator.
+     */
+    public Comparator<? super Tuple2<Double, Integer>> getQueueComparator() {
+        return comparator;
+    }
+
+    /**
+     * Gets dictionary data size.
+     *
+     * @return Dictionary data size.
+     */
+    public Integer getLength() {
+        return dictData.size();
+    }
+
+    public List<Tuple3<DenseMatrix, DenseVector, int[]>> getDictData() {
+        return dictData;
+    }
+
+    /** Encoder for the Knn model data. */
+    public static class ModelDataEncoder implements Encoder<Row> {
+        @Override
+        public void encode(Row modelData, OutputStream outputStream) throws IOException {
+            DataOutputView dataOutputView = new DataOutputViewStreamWrapper(outputStream);
+
+            DenseMatrixSerializer matrixSerializer = new DenseMatrixSerializer();
+            matrixSerializer.serialize((DenseMatrix) modelData.getField(0), dataOutputView);
+
+            DenseVectorSerializer vectorSerializer = new DenseVectorSerializer();
+            vectorSerializer.serialize((DenseVector) modelData.getField(1), dataOutputView);
+
+            int[] label = (int[]) Objects.requireNonNull(modelData.getField(2));
+            for (Integer integer : label) {
+                dataOutputView.writeInt(integer);
+            }
+        }
+    }
+
+    /** Decoder for the Knn model data. */
+    public static class ModelDataStreamFormat extends SimpleStreamFormat<Row> {
+
+        @Override
+        public Reader<Row> createReader(Configuration config, FSDataInputStream stream) {
+            return new Reader<Row>() {
+
+                @Override
+                public Row read() throws IOException {
+                    try {
+                        DataInputView source = new DataInputViewStreamWrapper(stream);
```

Comment:
Could we make `source`, `DenseMatrixSerializer()` and `DenseVectorSerializer()` private final fields of the anonymous `Reader` class, to avoid creating these instances repeated for every invocation of read()?

Reply:
OK